### PR TITLE
Compose Optics Across Different Cases ('matchOptional' & 'matchLens')

### LINKED ISSUE #132: Compose Optics Across Different Cases ('matchOptional' feature request)
Optics for a shared prop is easy:

```ts
import * as Op from 'monocle-ts/lib/Optional'

const seatsOptional: Op.Optional<Vehicle, number> = pipe(
  Op.id<Vehicle>(),
  Op.filter(Vehicle.isAnyOf(['Motorbike', 'Car'])),
  Op.prop('seats'),
);
```

(Or with monocle 2.2 & earlier):

```ts
const seatsOptional: Op.Optional<Vehicle, number> = Vehicle
  .prismFromPredicate(
    Vehicle.isAnyOf(['Motorbike', 'Car'])
  ).composeLens(
    Vehicle
      .select(['Motorbike', 'Car'])
      .lensFromProp('seats')
  )
```

But what if `seats` is unevenly nested across cases?

```ts
interface Car {
  type: 'Car'
  kind: 'electric' | 'fuel' | 'gaz'
  attributes: {
    power: number
    seats: number
  }
}
```

[Current best solution](https://gist.github.com/anthonyjoeseph/314b8f623b7f672524be38b070d5d888#file-seatoptional-ts-L45)

Proposed new solution:

```ts
import * as L from 'monocle-ts/lib/Lens'

const seatsOptional: Op.Optional<Vehicle, number> = Vehicle.matchOptional<number>({
  Motorbike: pipe(
    L.id<Motorbike>(),
    L.prop('seats'),
    L.asOptional,
  ),
  Car: pipe(
    L.id<Car>(),
    L.prop('attributes'),
    L.prop('seats'),
    L.asOptional,
  ),
  // undesired cases can be omitted
})
```

### PR DESCRIPTION
**Pull request by anthonyjoeseph**
Closes #132

### CODE REVIEW

**Comment by anthonyjoeseph** [comment]
related to haskell's [Total](http://hackage.haskell.org/package/total-1.0.5/docs/Lens-Family-Total.html) ([source](https://stackoverflow.com/questions/52832649/how-can-i-write-a-lens-for-a-sum-type))

File: packages/morphic-adt/src/matcher.ts [hunk 1]
```diff
@@ -98,6 +120,16 @@ export interface Matchers<A, Tag extends keyof A> {
...
   const match = (match: any, def?: any) => (a: any): any => (match[a[tag]] || def)(a)
+  const matchLens = (cases: any) =>
+    new Lens(
+      (s: any) => cases[s[tag]].get(s),
+      (a: any) => (s: any) => cases[s[tag]].set(a)(s)
+    )
+  const matchOptional = (cases: any) =>
+    new Optional(
+      (s: any) => (s[tag] in cases ? cases[s[tag]].getOption(s) : none),
+      (a: any) => (s: any) => (s[tag] in cases ? cases[s[tag]].set(a)(s) : s)
```
Comment (sledorze) [inline packages/morphic-adt/src/matcher.ts:131]: please, factorise the s[tag] access 
also reduce the cost of `s[tag] in cases` and `cases[s[tag]]` by just having a single access

**Comment by anthonyjoeseph** [comment]
Thank you, sure thing! Is this what you meant by 'factorise'?

**Comment by sledorze** [comment]
Yes better, we should remove the  'in' test and just check for undefined in the access result  (mapOptional)

**Comment by sledorze** [comment]
will be released soon as part of next alpha

**Comment by anthonyjoeseph** [comment]
Woohoo!  Thank you

**Comment by sledorze** [comment]
thanks @anthonyjoeseph for contributing! :)

**Pull request merged**